packages/frames.js/src/core/middleware/parseFramesMessage.test.ts

```diff
@@ -0,0 +1,111 @@
+import nock from "nock";
+import { redirect } from "../redirect";
+import { FramesContext } from "../types";
+import { parseFramesMessage } from "./parseFramesMessage";
+import { Message } from "@farcaster/core";
+
+describe("parseFramesMessage middleware", () => {
+  beforeEach(() => {
+    // make sure we don't introduce unexpected network requests
+    nock.disableNetConnect();
+  });
+
+  afterEach(() => {
+    nock.enableNetConnect();
+  });
+
+  it("moves to next middleware without parsing if request is not POST request", async () => {
+    const context: FramesContext = {
+      request: new Request("https://example.com"),
+    } as any;
+
+    const mw = parseFramesMessage();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("moves to next middleware if request is POST but does not have a valid JSON body", async () => {
+    const context: FramesContext = {
+      request: new Request("https://example.com", {
+        method: "POST",
+        body: "invalid json",
+      }),
+    } as any;
+
+    const mw = parseFramesMessage();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("moves to next middleware if request is POST with valid JSON but invalid body shape", async () => {
+    const context: FramesContext = {
+      request: new Request("https://example.com", {
+        method: "POST",
+        body: JSON.stringify({}),
+      }),
+    } as any;
+
+    const mw = parseFramesMessage();
+    const response = redirect("http://test.com");
+    const next = jest.fn(() => Promise.resolve(response));
+
+    await expect(mw(context, next)).resolves.toMatchObject(response);
+    expect(next).toHaveBeenCalledWith();
+  });
+
+  it("parses frame message from request body and adds it to context", async () => {
+    const context: FramesContext = {
+      request: new Request("https://example.com", {
+        method: "POST",
+        body: JSON.stringify({
+          trustedData: {
+            messageBytes: Buffer.from(
+              Message.encode(
+                Message.create({
+                  data: {
+                    fid: 123,
+                    frameActionBody: {
+                      castId: {
+                        fid: 456,
+                      },
+                      address: Buffer.from([0x789]),
+                      buttonIndex: 1,
+                      inputText: Buffer.from("hello"),
+                      state: Buffer.from(JSON.stringify({ test: true })),
```

Comment:
@stephancill what is a data shape in state? At the moment I'm just using the this and also in `renderResponse` middleware I'm serializing the state to `JSON`, so these two are compatible.

Does it make sense that state is just any valid JSON?

Reply:
from the spec:

> Frame servers are responsible for interpreting state values and may use whatever serialization format they wish

i think it is fair for frames.js to use JSON

Reply:
@stephancill ok so since I treat the value as JSON in `renderResponse`, I will implement automatic decoding here.

Reply:
I added automatic decoding.